### PR TITLE
AI Objective Panels-An informational Catastrophe.

### PR DESCRIPTION
**Pull request by TerryTehCoder**
## About The Pull Request
Adds non-functional Waypoint code to cyborgs. God Damn you Bot Code.

Introduces a new A.I Verb, "Cyborg Management Panel", with a full UI detailing Linked Cyborgs, their Charge, Location, and allows the A.I to send Private messages.

Also introduces a Objective crossreference system for A.I's and Cyborgs. Cyborgs are marked as either Occupied or Unoccupied on the A.I's Management panel(also found on their own Status tabs). Cyborgs can manually switch their readiness for tasking with the "Toggle Flag" verb.

A.I's can set an objective for any linked Cyborgs from their Management Panel. Cyborgs can always find their current assigned objective in their status panel. Alternatively, Cyborgs are free to set their own objectives with the "Set Current Objective" verb.

## Why It's Good For The Game

I've heard several agreements in the past about Cyborg Management and Tasking in General, i'm hoping to resolve that but having a headache in the process with BotCode. This gives the A.I another tool for managing large crowds of Cyborgs, and allows them to keep track of their synthetics even if they've switched off their Status Tab. 

Development Status: 

- [X] A.I Task/Objective Management
- [X] A.I Cyborg Management Panel
- [x] Cyborg Task System(bugs and errors)

### CODE REVIEW

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 1]
```diff
@@ -36,6 +36,18 @@
...
 	var/atom/movable/screen/robot_modules_background
+/** Cyborg pathing stuff-Unfinished
+	var/datum/atom_hud/data/bot_path/path_hud = new /datum/atom_hud/data/bot_path()
+	var/data_hud_type = DATA_HUD_DIAGNOSTIC_BASIC
+	var/path_image_icon = 'icons/mob/aibots.dmi'
+	var/path_image_icon_state = "path_indicator"
+	var/path_image_color = "#2189a3"
+	var/mob/living/silicon/ai/calling_ai //Important for identifying the A.I tasking it?
+	var/list/path = list() //Pathing for Cyborg Waypoints
+**/
```
Comment (Zonespace27) [inline code/modules/mob/living/silicon/robot/robot.dm:47]: Remove unneeded comment code please

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 2]
```diff
@@ -1262,3 +1277,92 @@
...
+		H.add_to_hud(src)
+
+/mob/living/silicon/robot/proc/flagwaypoint(caller, turf/waypoint) //Cyborgs deserve waypoints too! Poor saps..
+	set_path(get_path_to(src, waypoint, /turf/proc/Distance_cardinal, 0, 200))
+	calling_ai = caller
+	if(path && path.len) //Ensures that a valid path is calculated!
+		to_chat(caller, "<span class='danger'>Valid Route Calculated. Plotting...</span>")
+	else
+		to_chat(caller, "<span class='danger'>Failed to calculate a valid route. Ensure destination is clear of obstructions and within range.</span>")
+		set_path(null)
```
Comment (Zonespace27) [inline code/modules/mob/living/silicon/robot/robot.dm:1343]: ditto

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 3]
```diff
@@ -473,7 +480,27 @@
...
+		Cyborg = locate(href_list["cyborgmessage"]) in GLOB.alive_mob_list
+		if(!Cyborg)
+			return
+		to_chat(Cyborg, "<span class='notice'>Message recieved from stationbound A.I:<font color='yellow'> [cyborgmessage] </font></span>")
+/**	
+	if(href_list["flagwaypoint"])
+		Cyborg = locate(href_list["flagwaypoint"]) in GLOB.alive_mob_list
+		waypoint_mode = 1
+		if(!Cyborg)
+			return
```
Comment (Zonespace27) [inline code/modules/mob/living/silicon/ai/ai.dm:497]: ditto

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 4]
```diff
@@ -552,8 +579,10 @@
 		//The target must be in view of a camera or near the core.
 	if(turf_check in range(get_turf(src)))
 		call_bot(turf_check)
+//		call_cyborg()
 	else if(GLOB.cameranet && GLOB.cameranet.checkTurfVis(turf_check))
 		call_bot(turf_check)
+//		call_cyborg()
```
Comment (Zonespace27) [inline code/modules/mob/living/silicon/ai/ai.dm:585]: ditto

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 5]
```diff
@@ -610,6 +639,45 @@
...
+	c += "<table width='100%'><tr><td width='40%'><h3>Name</h3></td><td width='30%'><h3>Status</h3></td><td width='30%'><h3>Location</h3></td><td width='10%'><h3>Control</h3></td></tr>"
+
+	for(var/mob/living/silicon/robot/G in connected_robots)
+		if(G.get_virtual_z_level() == ai_Zlevel && !G.emagged) //Only non-emagged Cyborgs on the same Z-level are detected!
+			c += "<tr><td width='30%'>Designation: [G.name]</A> (Module: [G.designation])</td>"
+			c += "<td width='30%'>Integrity: [G.health]% 	 		Flagged as: [G.busystatus ? "Occupied" : "Unoccupied"]		Current Objective: [objectiveupdate]</td>"
+			c += "<td width='30%'>Location: [get_area_name(G, TRUE)]</td>" //Where is BORGIE????
+			c += "<td width='30%'>Cell: [G.cell ? "[G.cell.charge]/[G.cell.maxcharge]" : "Empty"]</td>"
+			c += "<td width='15%'><A HREF=?src=[REF(src)];cyborgmessage=[REF(G)]>Private Message</A></td>"
+//			c += "<td width='15%'><A HREF=?src=[REF(src)];flagwaypoint=[REF(G)]>Flag-Waypoint</A></td>" //Flag a Pain in the Ass is what this is.
```
Comment (Zonespace27) [inline code/modules/mob/living/silicon/ai/ai.dm:664]: ditto

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

**Comment by Zonespace27** [comment]
Tagging with DNM until I have a bit of time to properly review/test this

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 1 above)
Comment (TerryTehCoder) [reply code/modules/mob/living/silicon/robot/robot.dm]: Resolved.

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 6]
```diff
@@ -1262,3 +1268,27 @@
 		cell.charge = min(cell.charge + amount, cell.maxcharge)
 	if(repairs)
 		heal_bodypart_damage(repairs, repairs - 1)
+
+/mob/living/silicon/robot/proc/objectivesupdate()
+	var/mob/living/silicon/ai/AI
+	to_chat(src, "<span class='notice'>A.I Assigned Objective Updated. [AI.objectivesconfirm] Priority objective recieved as follows: [AI.objectiveupdate] </span>") //Objectives Updated, standby for Emergency Procedures. BEEEEP, BEEEP
+	busystatus = 1
+
+/mob/living/silicon/robot/verb/togglebusy()
```
Comment (Zonespace27) [inline code/modules/mob/living/silicon/robot/robot.dm:1277]: Give this some message to the cyborg. Currently, the cyborg gets no feedback but the ding.

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 7]
```diff
@@ -1262,3 +1268,27 @@
...
+	set category = "Robot Commands"
+	set name = "Toggle Flag"
+	set desc = "Toggle your busy flag off/online to indicate your status to your A.I"
+	busystatus = !busystatus
+	playsound(loc, 'sound/machines/ding.ogg', 25)
+	to_chat(usr, "<span class='notice>Flag Toggled [busystatus ? "Offline, unavailable for lowpriority tasking." : "Online, ready for assignment."]</span>")
+
```
Comment (Zonespace27) [inline code/modules/mob/living/silicon/robot/robot.dm:1284]: Add a verb for cyborgs to check their current objective + its priority

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 8]
```diff
@@ -610,6 +631,36 @@
...
+	var/turf/ai_current_turf = get_turf(src)
+	var/ai_Zlevel = ai_current_turf.get_virtual_z_level()
+	var/c
+	c += "<A HREF=?src=[REF(src)];cyborgrefresh=1>Query network status</A><br>"
+	c += "<table width='100%'><tr><td width='40%'><h3>Name</h3></td><td width='30%'><h3>Status</h3></td><td width='30%'><h3>Location</h3></td><td width='10%'><h3>Control</h3></td></tr>"
+
+	for(var/mob/living/silicon/robot/G in connected_robots)
+		if(G.get_virtual_z_level() == ai_Zlevel && !G.emagged) //Only non-emagged Cyborgs on the same Z-level are detected!
+			c += "<tr><td width='30%'>Designation: [G.name]</A> (Module: [G.designation])</td>"
+			c += "<td width='30%'>Integrity: [G.health]% 	 		Flagged as: [G.busystatus ? "Occupied" : "Unoccupied"]		Current Objective: [objectiveupdate]</td>"
```
Comment (Zonespace27) [inline code/modules/mob/living/silicon/ai/ai.dm:652]:
```suggestion
			c += "<td width='30%'>Integrity: [G.health]% 	 		
			<br>Flagged as: [G.busystatus ? "Occupied" : "Unoccupied"]		
			<br>Current Objective: [objectiveupdate]</td>"
```

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 7 above)
Comment (TerryTehCoder) [reply code/modules/mob/living/silicon/robot/robot.dm:1284]: Current objective is always available in the Cyborgs Status tab, and will always update without any need to refresh. Priority is not included in that status, i'll see about adding it.

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 6 above)
Comment (TerryTehCoder) [reply code/modules/mob/living/silicon/robot/robot.dm:1277]: A message was included in the original commit, but I missed a Semicolon that was preventing Line 1283 from reading. Fixed.

File: code/modules/mob/living/silicon/ai/ai.dm (hunk 8 above)
Comment (TerryTehCoder) [reply code/modules/mob/living/silicon/ai/ai.dm:652]: Suggestion gives off errors for unterminated text, and missing left-hand argument. I agree the status section of the UI is yucky. I'm open to alternative suggestions.

Comment (Zonespace27) [reply code/modules/mob/living/silicon/ai/ai.dm:652]: Ah, fucked up and didn't make the comment multiline

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 7 above)
Comment (TerryTehCoder) [reply code/modules/mob/living/silicon/robot/robot.dm:1285]: Current objective priority added to Status panel in 08b75d2.

File: code/modules/mob/living/silicon/ai/ai.dm (hunk 8 above)
Comment (TerryTehCoder) [reply code/modules/mob/living/silicon/ai/ai.dm:652]: That does the trick. It is a UI, so, it reads fine in the initial size, and if they want it drawn out, they can draw it out.

Comment (Zonespace27) [reply code/modules/mob/living/silicon/ai/ai.dm:652]: Sounds good!

**Comment by TerryTehCoder** [comment]
Might work on integrating the Set Current Objective verb in with the Cyborg HUD, but I'll save that for when someone with actual Sprite and Art capabilities comes about.

**Review by Zonespace27 — APPROVED** [review]

**Pull request merged**